### PR TITLE
health: Differentiate between unknown and unreachable state

### LINKED ISSUE #11874: Metrics are reporting unreachable nodes when Cilium starts up
The connectivity health checking metrics are reporting unreachable nodes in the beginning of the bootstrapping process. This should be avoided to avoid spikes in the metric when new nodes are being added.

Core issue: We don’t have a difference between “unknown” / “unreachable”.

### PR DESCRIPTION
**Pull request by vadorovsky**
Before this change, adding a new node with Cilium to the Kubernetes
cluster resulted in the new node being reported as unreachable in
cilium-health. That's because before Cilium gets fully provisions, the
node status is going to contain empty status (for HTTP probe).

That behavior was increasing the unreachable nodes Prometheus metric,
thus false alarming Cilium users and suggesting, that the provisioning
of the new node failed.

This change differentiates between unreachable and unhealthy status. In
order to be considered as unreachable, the status has to contain an
explicit error message, which is always set by the prober when it really
has problems with reaching already provisioned Cilium endpoint.
Otherwise its status is considered to be uknkown, which usually should
mean that the node is being provisioned.

Fixes #11874

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

```release-note
health: Differentiate between unknown and unreachable state in Cilium status
```

Outputs **Before the fix** when the 2nd node is provisioning:

```
vagrant@k8s1:~/go/src/github.com/cilium/cilium$ cilium metrics list | grep nodes
cilium_nodes_all_datapath_validations_total                                         37.000000
cilium_nodes_all_events_received_total        eventType="add" source="kvstore"      1.000000
cilium_nodes_all_events_received_total        eventType="add" source="local"        1.000000
cilium_nodes_all_events_received_total        source="kvstore" eventType="update"   4.000000
cilium_nodes_all_num                                                                2.000000
cilium_unreachable_nodes                                                            1.000000
vagrant@k8s1:~/go/src/github.com/cilium/cilium$ cilium-health status --succinct
Cluster health:      1/2 reachable   (2020-07-16T11:12:45Z)
  Name               IP              Reachable   Endpoints reachable
  k8s1 (localhost)   192.168.33.11   true        true
  k8s2               192.168.33.12   false       false
```

Outputs **after the fix** when the 2nd node is provisioning:

```
vagrant@k8s1:~/go/src/github.com/cilium/cilium$ cilium metrics list | grep nodes
cilium_nodes_all_datapath_validations_total                                         43.000000
cilium_nodes_all_events_received_total        eventType="add" source="kvstore"      1.000000
cilium_nodes_all_events_received_total        eventType="add" source="local"        1.000000
cilium_nodes_all_events_received_total        source="kvstore" eventType="update"   4.000000
cilium_nodes_all_num                                                                2.000000
cilium_unreachable_nodes                                                            0.000000
vagrant@k8s1:~/go/src/github.com/cilium/cilium$ cilium-health status --succinct
Cluster health:      2/2 reachable   (2020-07-16T11:48:37Z)
  Name               IP              Node        Endpoints
  k8s1 (localhost)   192.168.33.11   reachable   reachable
  k8s2               192.168.33.12   unknown     unknown
```

### CODE REVIEW

**Comment by mrostecki** [comment]
test-me-please

**Comment by mrostecki** [comment]
test-me-please

**Comment by mrostecki** [comment]
test-focus K8sPolicyTest

File: cilium/cmd/metrics_list.go [hunk 1]
```diff
@@ -48,5 +64,6 @@ var MetricsListCmd = &cobra.Command{
 
 func init() {
 	metricsCmd.AddCommand(MetricsListCmd)
+	MetricsListCmd.Flags().StringVarP(&matchPattern, "matchpattern", "p", "", "Show only metrics whose names match matchpattern")
```
Comment (aanm) [inline cilium/cmd/metrics_list.go]: `match-pattern`?

File: cilium/cmd/metrics_list.go [hunk 2]
```diff
@@ -3,13 +3,17 @@ package cmd
 import (
```
Comment (aanm) [inline cilium/cmd/metrics_list.go]: This commit looks like a new feature so it should be added in a different PR to avoid being backported to older branches.

File: Documentation/cmdref/cilium_metrics_list.md [hunk 3]
```diff
@@ -15,8 +15,9 @@ cilium metrics list [flags]
 ### Options
 
 ```
-  -h, --help            help for list
-  -o, --output string   json| jsonpath='{}'
+  -h, --help                  help for list
```
Comment (aanm) [inline Documentation/cmdref/cilium_metrics_list.md]: Should be squashed with previous commit

File: contrib/vagrant/start.sh [hunk 4]
```diff
@@ -595,11 +595,11 @@ if [ -n "${NO_PROVISION}" ]; then
     PROVISION_ARGS="--no-provision"
 fi
 if [ -n "${RELOAD}" ]; then
-    vagrant reload $PROVISION_ARGS
+    vagrant reload $PROVISION_ARGS $1
```
Comment (aanm) [inline contrib/vagrant/start.sh]: Should be done as part of a different PR

File: pkg/health/client/client.go [hunk 5]
```diff
@@ -237,7 +293,7 @@ func FormatHealthStatusResponse(w io.Writer, sr *models.HealthStatusResponse, pr
 		fmt.Fprintf(w, "Cluster health:\t%d/%d reachable\t(%s)\n",
 			healthy, len(sr.Nodes), sr.Timestamp)
 		if printAll || healthy < len(sr.Nodes) {
-			fmt.Fprintf(w, "  Name\tIP\tReachable\tEndpoints reachable\n")
+			fmt.Fprintf(w, "  Name\tIP\tNode\tEndpoints\n")
```
Comment (aanm) [inline pkg/health/client/client.go:297]: I think this will give an odd perception to the user:
```
$ cilium-health status --succinct
Cluster health:      2/2 reachable   (2020-07-16T11:48:37Z)
  Name               IP              Node        Endpoints
  k8s1 (localhost)   192.168.33.11   healthy     healthy
  k8s2               192.168.33.12   unknown     unknown
```

`cilium-health status` will only present the reachability of the node from the PoV of the node doing the probing. Presenting "healthy" / "unknown" / "unreachable" in the "Node" column might not be 100% clear. Instead of `healthy` I would suggest to use `reachable` instead.

File: pkg/metrics/status.go [hunk 6]
```diff
@@ -139,11 +139,15 @@ func (s *statusCollector) Collect(ch chan<- prometheus.Metric) {
 	)
```
Comment (aanm) [inline pkg/metrics/status.go:139]: Typo in the commit message and title "s/`Differnetiate`/`Differentiate`"

**Review by aanm — CHANGES_REQUESTED** [review]

**Comment by aanm** [comment]
Also, I think we can drop the backport to 1.6 and 1.7 since those older releases should receive more severe bug fixes instead.

**Comment by mrostecki** [comment]
@aanm Thanks! I created separate PRs - #12591 and #12590.

**Comment by mrostecki** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

File: pkg/health/client/client.go [hunk 7]
```diff
@@ -111,13 +128,51 @@ func Hint(err error) error {
...
+	}
+	// Initially assume healthy status.
+	status := ConnStatusReachable
+	for _, cs := range statuses {
+		switch GetConnectivityStatusType(cs) {
+		case ConnStatusUnreachable:
+			// If any status is unreachable, return it immediately.
+			return ConnStatusUnreachable
+		case ConnStatusUnknown:
+			// If the status is uknown, prepare to return it. It's
```
Comment (gandro) [inline pkg/health/client/client.go]: nit:
```suggestion
			// If the status is unknown, prepare to return it. It's
```

File: pkg/health/client/client.go [hunk 8]
```diff
@@ -154,7 +209,8 @@ func PathIsHealthy(cp *models.PathStatus) bool {
 		cp.HTTP,
 	}
 	for _, status := range statuses {
-		if !connectivityStatusHealthy(status) {
+		switch GetConnectivityStatusType(status) {
+		case ConnStatusUnreachable:
```
Comment (gandro) [inline pkg/health/client/client.go:214]: nit: I find it a bit weird that a function named `PathIsHealthy` now returns `true` if the state is actually unknown. The new logic and implementation itself is fine, just the old name does not match its new purpose.

I would prefer it if we renamed it to `pathIsUnreachable` or something (of course also invert the output). We can also make it private, as it seems unused outside of the package now.

Similarly, we could also rename `nodeIsHealthy` to `nodeIsUnreachable`.

File: pkg/health/client/client.go [hunk 9]
```diff
@@ -33,10 +33,27 @@ import (
 	"github.com/go-openapi/strfmt"
 )
 
+type ConnectivityStatusType int
```
Comment (gandro) [inline pkg/health/client/client.go:36]: nit: The description of this commit contains a typo: `uknkown`

**Review by gandro — APPROVED** [review]
Nice! Some very minor nits, they don't have to be addressed.

File: pkg/health/client/client.go (hunk 8 above)
Comment (vadorovsky) [reply pkg/health/client/client.go:214]: Yes, the name of that function might be misleading. I will change it

I also don't like the fact that it's not very clear what connectivity status do we have from the API model. We have the `Status` string value, we used to assume the non-healthy status if the string is non-empty. I find it quite misleading too. At the beginning I was thinking if I should start fixing that bug with changing the API structure, but then I decided not to do it, because it might be too intrusive for backporting.

But maybe changing the API would be still good as a separate refactoring PR just for 1.9.

Comment (vadorovsky) [reply pkg/health/client/client.go:214]: I will change the function name to `pathIsHealthyOrUnknown` just to avoid confusing double negations like `if !pathIsUnreachable(cp)`. Is it OK for you?

Comment (gandro) [reply pkg/health/client/client.go:214]: `pathIsHealthyOrUnknown` sounds good to me! :+1: 

> But maybe changing the API would be still good as a separate refactoring PR just for 1.9.

Agreed!

File: pkg/health/client/client.go (hunk 9 above)
Comment (vadorovsky) [reply pkg/health/client/client.go:36]: Fixed, thanks!

**Comment by mrostecki** [comment]
test-me-please

**Comment by joestringer** [comment]
These tests have been changed since the most recent test run, need to trigger them manually:

* K8s-1.12-Kernel-netnext Expected — Waiting for status to be reported
Required
* K8s-1.19-kernel-4.9 Expected — Waiting for status to be reported
Required

**Comment by joestringer** [comment]
retest-net-next

**Comment by joestringer** [comment]
retest-4.9

**Review by joestringer — APPROVED** [review]
